adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java

```diff
@@ -195,80 +192,356 @@ public void run() {
+         * Parses the parameters of a challenge String which has had its scheme removed
+         * (from its prefix).
+         *
+         * @param challengeSansScheme The challenge String, minus the scheme.
+         * @return A Map of the keys/values in the parsed parameters.
+         * @throws ResourceAuthenticationChallengeException If a parsing error is encountered or
+         *                                                  the String is malformed.
+         */
+        private static Map<String, String> parseParams(String challengeSansScheme) throws ResourceAuthenticationChallengeException {
+            if (StringExtensions.isNullOrBlank(challengeSansScheme)) {
+                throw new ResourceAuthenticationChallengeException(AUTH_HEADER_INVALID_FORMAT);
+            }
+
+            // Split on unquoted commas
+            final Map<String, String> params = new HashMap<>();
+            final String[] splitOnUnquotedCommas = challengeSansScheme.split(REGEX_SPLIT_UNQUOTED_COMMA, -1);
+            for (final String paramSet : splitOnUnquotedCommas) {
+                // Split keys/values by the '='
+                final String[] splitOnUnquotedEquals = paramSet.split(REGEX_SPLIT_UNQUOTED_EQUALS, -1);
+
+                // We should now have a left-side and right-side
+                if (splitOnUnquotedEquals.length != 2) {
+                    // Is this really what you want?
+                    throw new ResourceAuthenticationChallengeException(AUTH_HEADER_INVALID_FORMAT);
+                    //continue; // If there's no key/value pair, skip this token
                 }
 
-                String authority = headerItems.get(AUTHORITY_KEY);
-                if (!StringExtensions.isNullOrBlank(authority)) {
-                    authParams = new AuthenticationParameters(
-                            StringExtensions.removeQuoteInHeaderValue(authority),
-                            StringExtensions.removeQuoteInHeaderValue(headerItems.get(RESOURCE_KEY)));
-                } else {
-                    // invalid format
-                    throw new ResourceAuthenticationChallengeException(AUTH_HEADER_MISSING_AUTHORITY);
+                // Create the keys/values, trimming off any bogus whitespace
+                final String key = splitOnUnquotedEquals[0].trim();
+                final String value = splitOnUnquotedEquals[1].trim();
+
+                // if there is already a mapping for this key, we've seen this value before
+                // and should log a warning that this header looks fishy....
+                if (params.containsKey(key)) {
+                    Logger.w(TAG,
+                            "Key/value pair list contains redundant key. ",
+                            "Redundant key: " + key,
+                            ADALError.DEVELOPER_BEARER_HEADER_MULTIPLE_ITEMS);
                 }
-            } else {
+
+                // Add the key/value to the Map
+                params.put(key, value);
```

Comment:
Investigate: Do these keys/values need to be passed through `StringExtensions.urlFormDecode(String)` ?

Reply:
My testing is showing that the in/out values are not changed - the existing unit tests don't account for this. Will leave as-is for now and be ready to adjust if partner testing yields any issues